cpp/ql/test/README.md

```diff
@@ -0,0 +1,33 @@
+# C/C++ CodeQL tests
+
+This document provides additional information about the C/C++ CodeQL Tests located in `cpp/ql/test`.  See [Contributing to CodeQL](CONTRIBUTING.md) for general information about contributing to this repository.
+
+The tests can be run through Visual Studio Code.  Advanced users may also use the `codeql test run` command.
+
+## Contributing to the tests
+
+We are keen to have unit tests for all of our QL code.
+
+Every query in `cpp/ql/src` (outside of `cpp/ql/src/experimental`) should have a test in the corresponding subdirectory of `cpp/ql/test/query-tests`. At a minimum, each query test shall contain one case that should be detected by the query, and one related case that should not.
+
+TODO: example?
+
+Library features should also have test coverage, in `cpp/ql/test/library-tests`.
+
+## Copying code
+
+The contents of `cpp/ql/test` should be original - nothing should be copied from other sources. In particular do not copy-paste C/C++ code from third party projects, your own projects, or the standard C/C++ library implementation of your compiler (regardless of the associated license). As an exception, required declarations may be taken from the following sources where necessary:
+ - ISO/IEC Programming languages - C (all versions)
+ - ISO/IEC Programming languages - C++ (all versions)
+
```

Comment:
We're also allowed to "translate QL to C++" (cc @p0). For example, from https://github.com/github/codeql/blob/89a4cff5f848536312493b76d4313e8ce792e943/cpp/ql/src/semmle/code/cpp/models/implementations/Strcpy.qll#L59-L67 we can reconstruct a declaration like

```c
void strcpy(char *dest, const char *src);
```

This declaration doesn't have the correct return type because the return type can't be guessed from the QL code. It doesn't matter that the return type is wrong for the same reason: it doesn't matter to the QL code.

Reply:
I think we're also allowed to use any code that's in the public domain. Maybe that goes without saying since that's the definition of the public domain.

Reply:
> We're also allowed to "translate QL to C++"

I don't recall the exact conclusion of that conversation, but if we agree this is allowed and useful, I'll add it.

> I think we're also allowed to use any code that's in the public domain.

That's sounds reasonable, though not strictly risk free IMO.  Again if we agree this is allowed, I'm happy to add it.

I'd be glad to hear more thoughts on these points, especially from someone with appropriate legal knowledge, authority, or experience from other teams to share.

Reply:
Approved: https://github.slack.com/archives/CP0LHP150/p1606469775045400?thread_ts=1606467776.045100&cid=CP0LHP150

Reply:
I've added both.  I'm not sure whether it's worth adding an example of translating QL to C++ as well, or whether this will just distract from the core message.